main/src/ca/uwaterloo/flix/language/phase/EffectBinder.scala

```diff
@@ -123,69 +129,160 @@ object EffectBinder {
+      val e2 = letBindEffectsTopLevel(exp2)
+      val e3 = letBindEffectsTopLevel(exp3)
+      Expr.IfThenElse(e1, e2, e3, tpe, purity, loc)
 
     case Expr.Branch(exp, branches, tpe, purity, loc) =>
-      Expr.Branch(exp, branches, tpe, purity, loc)
+      val e = letBindEffectsTopLevel(exp)
+      val bs = branches.map {
+        case (sym, branchExp) => (sym, letBindEffectsTopLevel(branchExp))
+      }
+      Expr.Branch(e, bs, tpe, purity, loc)
 
     case Expr.JumpTo(sym, tpe, purity, loc) =>
       Expr.JumpTo(sym, tpe, purity, loc)
 
-    case Expr.Let(sym, exp1, exp2, tpe, purity, loc) =>
-      Expr.Let(sym, exp1, exp2, tpe, purity, loc)
+    case Expr.Let(sym, exp1, exp2, _, _, loc) =>
+      val e1 = letBindInnerEffects(binders)(exp1)
+      binders.addOne(Expr.Let(sym, e1, null, null, null, loc))
+      letBindInnerEffects(binders)(exp2)
 
-    case Expr.LetRec(varSym, index, defSym, exp1, exp2, purity, tpe, loc) =>
-      Expr.LetRec(varSym, index, defSym, exp1, exp2, purity, tpe, loc)
+    case Expr.LetRec(varSym, index, defSym, exp1, exp2, _, _, loc) =>
+      val e1 = letBindInnerEffects(binders)(exp1)
+      binders.addOne(Expr.LetRec(varSym, index, defSym, e1, null, null, null, loc))
+      letBindInnerEffects(binders)(exp2)
 
     case Expr.Scope(sym, exp, tpe, purity, loc) =>
-      Expr.Scope(sym, exp, tpe, purity, loc)
+      val e = letBindEffectsTopLevel(exp)
+      Expr.Scope(sym, e, tpe, purity, loc)
 
     case Expr.TryCatch(exp, rules, tpe, purity, loc) =>
-      Expr.TryCatch(exp, rules, tpe, purity, loc)
+      val e = letBindEffectsTopLevel(exp)
+      Expr.TryCatch(e, rules, tpe, purity, loc)
 
     case Expr.TryWith(exp, effUse, rules, tpe, purity, loc) =>
-      Expr.TryWith(exp, effUse, rules, tpe, purity, loc)
+      val e = letBindEffectsTopLevel(exp)
+      val rs = rules.map {
+        case HandlerRule(op, fparams, handlerExp) => HandlerRule(op, fparams, letBindEffectsTopLevel(handlerExp))
+      }
+      Expr.TryWith(e, effUse, rs, tpe, purity, loc)
 
     case Expr.Do(op, exps, tpe, purity, loc) =>
-      Expr.Do(op, exps, tpe, purity, loc)
+      val es = exps.map(letBindEffects(binders))
+      Expr.Do(op, es, tpe, purity, loc)
 
     case Expr.NewObject(name, clazz, tpe, purity, methods, exps, loc) =>
-      Expr.NewObject(name, clazz, tpe, purity, methods, exps, loc)
+      val es = exps.map(letBindInnerEffects(binders))
+      Expr.NewObject(name, clazz, tpe, purity, methods, es, loc)
 
     case Expr.Resume(exp, tpe, loc) =>
-      Expr.Resume(exp, tpe, loc)
+      val e = letBindEffects(binders)(exp)
+      Expr.Resume(e, tpe, loc)
+  }
+
+  /**
+    * Let-binds all sub-expressions including the top-most one.
+    * Always returns a variable or a constant.
+    */
+  private def letBindEffects(binders: mutable.ArrayBuffer[Expr])(exp: Expr)(implicit lctx: LocalContext, flix: Flix): Expr = {
+    val expTransformed = letBindInnerEffects(binders)(exp)
```

Comment:
This variable is used in the call on line 277, so it would be meaningful to move it there to make clear both what 277 does and also that thebind function does not close over it.